### PR TITLE
Make cropping centre an argument

### LINKED ISSUE #86: No way to specify centre of image when cropping

### PR DESCRIPTION
**Pull request by thomasmfish**
Fixes #86

### CODE REVIEW

File: src/sim_recon/images/utils.py [hunk 1]
```diff
@@ -12,33 +14,94 @@
...
 
+def _calculate_shape_crop_slices(
+    target_xy_shape: NDArray[np.integer[Any]],
+    current_xy_shape: NDArray[np.integer[Any]],
+    centre: NDArray[np.integer[Any]],
+) -> tuple[slice, slice]:
+    if np.any(target_xy_shape > current_xy_shape):
+        raise OutOfBoundsError("Target crop shape exceeds at least one array dimension")
+
+    from_centre = (target_xy_shape - 1) / 2
```
Comment (PeterC-DLS) [inline src/sim_recon/images/utils.py]: Better to use `(target_xy_shape - 1)// 2` and remove `ceil` calls?

File: src/sim_recon/images/utils.py [hunk 2]
```diff
@@ -12,33 +14,94 @@
...
     xy_shape: tuple[int, int] | None = None,
     crop: float = 0,
+    xy_centre: tuple[int, int] | None = None,
 ) -> NDArray[Any]:
+
     assert array.ndim > 1, "Images must have > 1 dimension"
+
     array_slices = [slice(None)] * array.ndim
+
+    current_xy_shape = np.asarray((array.shape[-1], array.shape[-2]), dtype=np.uint16)
```
Comment (PeterC-DLS) [inline src/sim_recon/images/utils.py]: Seems like your code is contorted by preserving the x-y ordering rather transposing to the storage ordering

File: src/sim_recon/images/utils.py [hunk 3]
```diff
@@ -12,33 +14,94 @@
 logger = logging.getLogger(__name__)
 
 
+def _calculate_shape_crop_slices(
+    target_xy_shape: NDArray[np.integer[Any]],
```
Comment (PeterC-DLS) [inline src/sim_recon/images/utils.py]: Function is agnostic to whether it's `xy` or `yx`

File: src/sim_recon/images/utils.py (hunk 1 above)
Comment (thomasmfish) [reply src/sim_recon/images/utils.py]: `np.ceil` rounds up for `max_bounds` and down for `min_bounds`, so that doesn't work

**Comment by thomasmfish** [comment]
I found some issues in the process of making those changes, so I've added more tests and changed how it works a bit to fix those issues.

**Comment by thomasmfish** [comment]
Just updated to allow float coordinates for `--centre`, since that is possible.

File: src/sim_recon/images/utils.py [hunk 4]
```diff
@@ -12,33 +14,107 @@
...
+    )
+
+
+def _calculate_2d_decimal_crop_slices(
+    crop: float,
+    current_shape: NDArray[np.integer[Any]],
+    centre: NDArray[np.floating[Any]],
+) -> tuple[slice, slice]:
+    if crop <= 0 or crop >= 1:
+        raise InvalidValueError("Argument 'crop' must be >= 0 and < 1")
```
Comment (PeterC-DLS) [inline src/sim_recon/images/utils.py]: Either `crop < 0` in test or `> 0` in error message

File: tests/test_images/test_utils.py [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+            ).astype(np.uint16)
+            assert_array_equal(
+                result_2d_shape, expected_shape, err_msg="expected shape does not match"
+            )
+        assert_array_equal(
+            result, expected_output, err_msg="result doesn't match expected output"
+        )
+
+
+def test_complex_to_interleaved_float(): ...
```
Comment (PeterC-DLS) [inline tests/test_images/test_utils.py]: Unfinished tests?

Comment (thomasmfish) [reply tests/test_images/test_utils.py]: Ah yeah, removed them as they wouldn't fit in the PR anyway

**Review by PeterC-DLS — APPROVED** [review]
LGTM

**Pull request merged**